### PR TITLE
Separate state crate

### PR DESCRIPTION
**Pull request by lee-orr**
# Objective

Extracts the state mechanisms into a new crate called "bevy_state".

This comes with a few goals:

- state wasn't really an inherent machinery of the ecs system, and so keeping it within bevy_ecs felt forced
- by mixing it in with bevy_ecs, the maintainability of our more robust state system was significantly compromised

moving state into a new crate makes it easier to encapsulate as it's own feature, and easier to read and understand since it's no longer a single, massive file.

## Solution

move the state-related elements from bevy_ecs to a new crate

## Testing

- Did you test these changes? If so, how? all the automated tests migrated and passed, ran the pre-existing examples without changes to validate.

---

## Migration Guide

Since bevy_state is now gated behind the `bevy_state` feature, projects that use state but don't use the `default-features` will need to add that feature flag.

Since it is no longer part of bevy_ecs, projects that use bevy_ecs directly will need to manually pull in `bevy_state`,  trigger the StateTransition schedule, and handle any of the elements that bevy_app currently sets up.

### CODE REVIEW

File: examples/ecs/sub_states.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@
 //! a substate called `IsPaused` to track whether the game is paused or not.
 
 use bevy::prelude::*;
+use bevy::state::state::StateTransitionEvent;
```
Comment (MiniaczQ) [inline examples/ecs/sub_states.rs:11]: This should be in the prelude when using states

**Review by MiniaczQ — COMMENTED** [review]
Looks good, much cleaner now that it's in separate files.

Proposal: move state related examples to separate folder?

**Comment by alice-i-cecile** [comment]
Yeah, we should move the examples to a separate folder now :)

File: crates/bevy_state/macros/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+edition = "2021"
+license = "MIT OR Apache-2.0"
+
+[lib]
+proc-macro = true
+
+[dependencies]
+bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.14.0-dev" }
+
+syn = { version = "2.0", features = ["full"] }
```
[the remark this answers is not in the document]
Comment (lee-orr) [reply crates/bevy_state/macros/Cargo.toml:14]: I'd forgotten about that! Good catch

**Review by MiniaczQ — APPROVED** [review]

**Review by alice-i-cecile — COMMENTED** [review]
You need to update the publish.sh script, but otherwise this LGTM.

https://github.com/bevyengine/bevy/blob/main/tools/publish.sh

**Comment by lee-orr** [comment]
> You need to update the publish.sh script, but otherwise this LGTM.
> 
> https://github.com/bevyengine/bevy/blob/main/tools/publish.sh

Done!

**Review by MiniaczQ — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

File: tools/publish.sh [hunk 3]
```diff
@@ -11,6 +11,7 @@ crates=(
     bevy_reflect
     bevy_ecs/macros
     bevy_ecs
+    bevy_state
```
Comment (BD103) [inline tools/publish.sh:14]: `bevy_state_macros` will need to be added as well.

**Review by BD103 — CHANGES_REQUESTED** [review]
Just one quick thing before this gets merged!

File: tools/publish.sh (hunk 3 above)
Comment (orbachl-mcmaster) [reply tools/publish.sh:14]: Thanks! I can't believe I forgot that

Comment (lee-orr) [reply tools/publish.sh:15]: Thanks! I can't believe I forgot that!

**Review by BD103 — APPROVED** [review]

File: crates/bevy_ecs/macros/Cargo.toml [hunk 4]
```diff
@@ -11,7 +11,7 @@ proc-macro = true
 [dependencies]
 bevy_macro_utils = { path = "../../bevy_macro_utils", version = "0.14.0-dev" }
 
-syn = { version = "2.0", features = ["full"] }
+syn = "2.0"
```
Comment (kristoff3r) [inline crates/bevy_ecs/macros/Cargo.toml:14]: This is a bad merge from main where the same thing was fixed. I think the merge might go through if this change is removed.

**Review by MiniaczQ — APPROVED** [review]

**Pull request merged**